File: week09/5/prob2.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        // color : curr 노드에 칠해진 색깔 (1 or 2)
        // - 0 : unvisited, not-colored
        vector<int> colors(graph.size(), 0);

        for (int source=0; source<graph.size(); source++) {
            if (colors[source] == 0) {
                if (!dfs_helper(source, 1, graph, colors)) {
                    return false;
                }
            }
        }
        return true;
    }

    bool dfs_helper(int curr, int color, const vector<vector<int>>& graph, vector<int>& colors) {
        colors[curr] = color;

        for (int neighbor : graph[curr]) {
            // base case : 잘못됨 감지
            if (colors[neighbor] == color) {
                return false;
            }
            // recursive case : unvisited neighbor
            else if (colors[neighbor] == 0) {
                // 3 - color : 1 -> 2 or 2 -> 1
                if (!dfs_helper(neighbor, 3 - color, graph, colors)) {
                    return false;
                }
            }
        }
        return true;
    }
};

File: week09/2/Prob2.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        # initialize color array, -1: not colored, 0 or 1: colored
        colors = [-1] * n

        # use BFS method
        for i in range(n):
            if colors[i] == -1:
                queue = deque([i])
                colors[i] = 0
                
                while queue:
                    node = queue.popleft()
                    current_color = colors[node]
                    next_color = 1 - current_color # 현재 노드의 색과 다른 색으로 설정: 1이면 0, 0이면 1로 처리

                    for neighbor in graph[node]:
                        if colors[neighbor] == -1:
                            colors[neighbor] = next_color
                            queue.append(neighbor)
                        elif colors[neighbor] == current_color:
                            return False
        return True
        